hathitrust/ht_indexer
# Failed messages are back to the queue using a dead-letter-exchange and capturing exceptions;

This PR aims to enhance exception handling and produce more efficient logs. It also applied the following changes:

* Added more unit tests; improved their readability; removed some duplicated code
* Upgrade Python packages
* Add qodana as a code analysis tool (For testing)

See below images with some of the information you will find in the logs

**document_retriever logs**

![image](https://github.com/hathitrust/ht_indexer/assets/47088146/4b3d11d9-68b2-40e1-b935-a4fa9790e248)
![image](https://github.com/hathitrust/ht_indexer/assets/47088146/25f89cec-1fce-44fb-bf2c-7fbe99eeebf8)
![image](https://github.com/hathitrust/ht_indexer/assets/47088146/0888ed31-5dc6-446a-95d1-f06f0e82d4fd)

**document_generator logs**
* well-generated document log
![image](https://github.com/hathitrust/ht_indexer/assets/47088146/e4015814-b930-456d-a68a-946cbc807d5b)

* failed generated document log
![image](https://github.com/hathitrust/ht_indexer/assets/47088146/cc140d2f-98d8-438a-a79b-1774533f5e0d)

**document_indexer logs**
![image](https://github.com/hathitrust/ht_indexer/assets/47088146/61f458e4-3bc3-423c-86cc-037d3f4ddbe0)

Additional information:
* The starting point of the current PR was this [PR](https://github.com/hathitrust/ht_indexer/pull/14), then here you will find all the changes. The current PR is up-to-date with the previous one.

* The way of testing it is the same as explained in the PR (https://github.com/hathitrust/ht_indexer/pull/14), but I will copy below how to test it.

* I have been working with two branches in parallel to test my skills using **git rebase** and also to measure the effort of building several features in different branches. 
      **Takeaways**: It makes sense to split work into different pull requests if they touch different files, otherwise solving conflicts can be challenging and time-consuming.

@aelkiss, please choose what option works better for you:
 **Option 1**: Review the [PR](https://github.com/hathitrust/ht_indexer/pull/14) and the current PR as an independent task
 **Option 2** Review the current PR as it contains all the changes on the [PR](https://github.com/hathitrust/ht_indexer/pull/14)

If you choose option 2, I will discard this [PR](https://github.com/hathitrust/ht_indexer/pull/14). Otherwise, we continue working with independent PRs.

**How to test it**:

Step 1. Create /sdr1/obj directory
mkdir ../sdr1/obj

Step 2. Retrieve from pairtree repository data for testing
export HT_SSH_HOST=dev-2.babel.hathitrust.org
scp $HT_SSH_HOST:/sdr1/obj/umn/pairtree_root/31/95/10/01/99/77/04/p/31951001997704p/31951001997704p{.zip,mets.xml} ../sdr1/obj

Step 3. Create the image
docker build -t document_generator .

Step 4. Run the container to retrieve documents

docker compose up document_retriever -d

Open [RabbitQM](http://localhost:15672/#/) and [Solr](http://localhost:8983/solr/#/core-x/query?q=*:*&q.op=OR&indent=true) to check the success of this tests

Step 5. Run retriever service

docker compose exec document_retriever python document_retriever_service/full_text_search_retriever_service.py --list_documents umn.31951001997704p,mdp.35112103801405 --query_field item

Checking: You should see 2 messages on the retriever_queue

Step 6. Run the container to generate documents

docker compose up document_generator

Checking: You should see the document umn.31951001997704p on indexer_queue and the document mdp.35112103801405 on the retriever_queue_dead_letter_queue. As you do not have the files associated with mdp.35112103801405, this message will be rejected.

Step 8. Run the container to index documents
docker compose up document_indexer

Checking: The document umn.31951001997704p should be indexed in [Solr](http://localhost:8983/solr/#/core-x/query?q=*:*&q.op=OR&indent=true) and the retriever_queue and indexer_queue are empty. However, you should find the document mdp.35112103801405 in the retriever_queue_dead_letter_queue.
